src/shogun/kernel/MaternKernel.cpp

```diff
@@ -0,0 +1,170 @@
+/*
+ * This software is distributed under BSD 3-clause license (see LICENSE file).
+ *
+ * Authors: Gil Hoben
+ */
+
+#include <shogun/distance/EuclideanDistance.h>
+#include <shogun/kernel/MaternKernel.h>
+
+using namespace shogun;
+
+
+MaternKernel::MaternKernel(): ShiftInvariantKernel()
+{
+	register_params();
+}
+
+MaternKernel::MaternKernel(int32_t size, float64_t w, float64_t order): 
+	ShiftInvariantKernel(), m_nu(order), m_width(w)
+{
+	register_params();
+	set_cache_size(size);
+}
+
+MaternKernel::MaternKernel(float64_t w, float64_t order): MaternKernel(10, w, order)
+{
+}
+
+MaternKernel::MaternKernel(const std::shared_ptr<Features>& l, const std::shared_ptr<Features>& r,
+	int32_t size, float64_t w, float64_t order): MaternKernel(size, w, order)
+{
+	init(l, r);
+}
+
+MaternKernel::~MaternKernel()
+{
+	cleanup();
+}
+
+void MaternKernel::cleanup()
+{
+	Kernel::cleanup();
+	m_distance->cleanup();
+}
+
+bool MaternKernel::init(std::shared_ptr<Features> l, std::shared_ptr<Features> r)
+{
+	cleanup();
+	return ShiftInvariantKernel::init(l,r);
+}
+
+void MaternKernel::register_params()
+{
+	auto distance=std::make_shared<EuclideanDistance>();
+	distance->set_disable_sqrt(false);
+	m_distance=distance;
+	SG_ADD(&m_nu, "nu", "order of the Bessel function of the second kind", ParameterProperties::HYPER)
+	SG_ADD(&m_width, "width", "Kernel scale", ParameterProperties::HYPER)
+}
+
+float64_t MaternKernel::compute(int32_t idx_a, int32_t idx_b)
+{
+	float64_t result;
+	const float64_t dist = ShiftInvariantKernel::distance(idx_a, idx_b);
+
+	// first we can check if we should use one of the approximations which are cheaper to calculate
+	// these formulas were taken from Rasmussen's GPs for ML
+	if (Math::fequals(m_nu, 0.5, std::numeric_limits<float64_t>::epsilon()))
+		result = std::exp(-dist/m_width);
+	else if (Math::fequals(m_nu, 1.5, std::numeric_limits<float64_t>::epsilon()))
+	{
+		float64_t ratio = (std::sqrt(3) * dist) / m_width;
```

Comment:
is something like sqrt(3) constexpr? :)

Reply:
I think all compilers would precompute this with optimisations on

Reply:
i see! cool!

Reply:
see line 2 of the assembly in https://godbolt.org/z/xCXF3q. When optimisations are turned off this becomes a runtime call to sqrt